src/VisualStudio/Core/Def/Implementation/ProjectSystem/DocumentProvider.cs

```diff
@@ -207,51 +331,78 @@ public void NotifyDocumentRegisteredToProject(IVisualStudioHostDocument document
             uint docCookie;
             if (RunningDocumentTable.TryGetCookieForInitializedDocument(document.Key.Moniker, out docCookie))
             {
-                TryProcessOpenForDocCookie(docCookie);
+                TryProcessOpenForDocCookie(docCookie, cancellationToken);
             }
+
+            cancellationToken.ThrowIfCancellationRequested();
+            RemovePendingDocumentInitializationTask(document);
         }
 
-        private void TryProcessOpenForDocCookie(uint docCookie)
+        private void TryProcessOpenForDocCookie(uint docCookie, CancellationToken cancellationToken)
         {
+            AssertIsForeground();
+
+            cancellationToken.ThrowIfCancellationRequested();
```

Comment:
I had to carefully place the `cancellationToken.ThrowIfCancellationRequested();` calls in this method to avoid corrupting consistency between data structures and I avoided putting it within the loops and I am still not sure if I have done the right thing. Probably better to take away the cancellation token for simplicity and we can add it later if perf necessitates it.